File: scripts/validate_file_sizes.py
#!/usr/bin/env python3
"""
File size validation script for server code quality refactoring.

Ensures no Python file exceeds the maximum line count threshold.
"""

import os
import sys
from pathlib import Path
from typing import List, Tuple

MAX_LINES = 500
SERVER_DIR = Path(__file__).parent.parent / "server"


def count_lines(file_path: Path) -> int:
    """Count non-empty lines in a file."""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            return sum(1 for line in f if line.strip())
    except Exception as e:
        print(f"Error reading {file_path}: {e}")
        return 0


def find_python_files(directory: Path) -> List[Path]:
    """Find all Python files in directory recursively."""
    return list(directory.rglob("*.py"))


def validate_file_sizes(max_lines: int = MAX_LINES) -> Tuple[List[Tuple[Path, int]], List[Tuple[Path, int]]]:
    """
    Validate all Python files are under the maximum line count.

    Returns:
        Tuple of (violations, compliant_files)
    """
    violations = []
    compliant = []

    python_files = find_python_files(SERVER_DIR)

    for file_path in python_files:
        # Skip test files and __pycache__
        if '__pycache__' in str(file_path) or 'test_' in file_path.name:
            continue

        line_count = count_lines(file_path)

        if line_count > max_lines:
            violations.append((file_path, line_count))
        else:
            compliant.append((file_path, line_count))

    return violations, compliant


def main():
    """Run file size validation."""
    print(f"Validating Python files in {SERVER_DIR}")
    print(f"Maximum allowed lines: {MAX_LINES}\n")

    violations, compliant = validate_file_sizes()

    # Sort by line count descending
    violations.sort(key=lambda x: x[1], reverse=True)
    compliant.sort(key=lambda x: x[1], reverse=True)

    if violations:
        print("=" * 60)
        print("VIOLATIONS (files exceeding limit):")
        print("=" * 60)
        for file_path, lines in violations:
            rel_path = file_path.relative_to(SERVER_DIR.parent)
            print(f"  {lines:4d} lines - {rel_path}")
        print()

    print("=" * 60)
    print(f"SUMMARY:")
    print("=" * 60)
    print(f"  Total files checked: {len(violations) + len(compliant)}")
    print(f"  Compliant files: {len(compliant)}")
    print(f"  Violations: {len(violations)}")

    if violations:
        print(f"\n[FAIL] {len(violations)} file(s) exceed {MAX_LINES} lines")
        sys.exit(1)
    else:
        print(f"\n[PASS] All files under {MAX_LINES} lines")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: scripts/validate-file-sizes.js
#!/usr/bin/env node

/**
 * Automated file size validation script for client code quality refactoring
 * Validates that no TypeScript file exceeds 500 lines of code
 * Requirements: 10.1, 10.5
 */

const fs = require('fs');
const path = require('path');
const glob = require('glob');

const MAX_LINES = 500;
const CLIENT_SRC_PATH = 'client/src';

/**
 * Count lines in a file, excluding empty lines and comments-only lines
 */
function countLinesInFile(filePath) {
    try {
        const content = fs.readFileSync(filePath, 'utf8');
        const lines = content.split('\n');
        
        // Count non-empty, non-comment-only lines
        let lineCount = 0;
        for (const line of lines) {
            const trimmed = line.trim();
            if (trimmed && !trimmed.startsWith('//') && !trimmed.startsWith('/*') && trimmed !== '*/') {
                lineCount++;
            }
        }
        
        return lineCount;
    } catch (error) {
        console.error(`Error reading file ${filePath}:`, error.message);
        return 0;
    }
}

/**
 * Get all TypeScript files in the client source directory
 */
function getAllTypeScriptFiles() {
    const pattern = 'client/src/**/*.ts';
    const options = { 
        ignore: ['**/*.test.ts', '**/*.spec.ts', '**/node_modules/**']
    };
    
    const files = glob.sync(pattern, options);
    console.log(`Found ${files.length} TypeScript files matching pattern: ${pattern}`);
    
    return files;
}

/**
 * Validate file sizes and report violations
 */
function validateFileSizes() {
    console.log('🔍 Validating TypeScript file sizes...\n');
    
    const tsFiles = getAllTypeScriptFiles();
    const violations = [];
    let totalFiles = 0;
    
    for (const filePath of tsFiles) {
        const lineCount = countLinesInFile(filePath);
        totalFiles++;
        
        if (lineCount > MAX_LINES) {
            violations.push({
                file: filePath,
                lines: lineCount,
                excess: lineCount - MAX_LINES
            });
        }
        
        // Show progress for large codebases
        if (totalFiles % 50 === 0) {
            console.log(`Processed ${totalFiles} files...`);
        }
    }
    
    // Report results
    console.log(`\n📊 Validation Results:`);
    console.log(`Total TypeScript files checked: ${totalFiles}`);
    console.log(`Maximum allowed lines: ${MAX_LINES}`);
    
    if (violations.length === 0) {
        console.log('✅ All files comply with size constraints!');
        return true;
    } else {
        console.log(`❌ Found ${violations.length} files exceeding size limit:\n`);
        
        // Sort by excess lines (worst first)
        violations.sort((a, b) => b.excess - a.excess);
        
        for (const violation of violations) {
            console.log(`  ${violation.file}`);
            console.log(`    Lines: ${violation.lines} (${violation.excess} over limit)`);
            console.log('');
        }
        
        console.log('🔧 These files need refactoring to meet size constraints.');
        return false;
    }
}

/**
 * Generate refactoring report with recommendations
 */
function generateRefactoringReport(violations) {
    const reportPath = 'refactoring-report.md';
    let report = '# File Size Refactoring Report\n\n';
    report += `Generated: ${new Date().toISOString()}\n\n`;
    report += `## Files Requiring Refactoring\n\n`;
    
    for (const violation of violations) {
        report += `### ${violation.file}\n`;
        report += `- **Current size**: ${violation.lines} lines\n`;
        report += `- **Excess**: ${violation.excess} lines over limit\n`;
        report += `- **Priority**: ${violation.excess > 1000 ? 'Critical' : violation.excess > 500 ? 'High' : 'Medium'}\n\n`;
    }
    
    fs.writeFileSync(reportPath, report);
    console.log(`📄 Detailed report saved to: ${reportPath}`);
}

// Main execution
if (require.main === module) {
    const isValid = validateFileSizes();
    process.exit(isValid ? 0 : 1);
}

module.exports = {
    validateFileSizes,
    countLinesInFile,
    getAllTypeScriptFiles,
    MAX_LINES
};